File: project/core/connector.py
from typing import Any, Callable
from loguru import logger

class Connector:
    def __init__(self, source_module: str, target_module: str):
        self.source = source_module
        self.target = target_module
        self.validators: list[Callable] = []
        logger.info(f"Connector created between {source_module} and {target_module}")

    def add_validator(self, validator: Callable[[Any], bool]):
        self.validators.append(validator)
        logger.debug(f"Validator added to connector {self.source}->{self.target}")

    def validate_data(self, data: Any) -> bool:
        for validator in self.validators:
            try:
                if not validator(data):
                    logger.warning(f"Validation failed for {self.source}->{self.target}")
                    return False
            except Exception as e:
                logger.error(f"Validator error: {e}")
                return False
        return True

    def transfer_data(self, data: Any) -> bool:
        if self.validate_data(data):
            logger.info(f"Data transferred from {self.source} to {self.target}")
            return True
        return False

File: src/core/connector.ts
import { logger } from './logger';

type Validator = (data: any) => boolean;

export class Connector {
  private validators: Validator[];

  constructor(
    private sourceModule: string,
    private targetModule: string
  ) {
    this.validators = [];
    logger.info({ source: sourceModule, target: targetModule }, 'Connector created');
  }

  addValidator(validator: Validator): void {
    this.validators.push(validator);
    logger.debug({ source: this.sourceModule, target: this.targetModule }, 'Validator added');
  }

  private validateData(data: any): boolean {
    return this.validators.every(validator => {
      try {
        return validator(data);
      } catch (error) {
        logger.error({ error, data }, 'Validator error');
        return false;
      }
    });
  }

  transferData(data: any): boolean {
    if (this.validateData(data)) {
      logger.info({ source: this.sourceModule, target: this.targetModule }, 'Data transferred');
      return true;
    }
    logger.warn({ source: this.sourceModule, target: this.targetModule, data }, 'Data transfer failed');
    return false;
  }
}
